lib/carbon/cache.py

```diff
@@ -137,11 +156,42 @@ def drain_metric(self):
     else:
       # Avoid .keys() as it dumps the whole list
       metric = self.iterkeys().next()
-    return (metric, self.pop(metric))
+
+    # Do not pop all datapoints for this metric.
+    # Instead, retain last N datapoints in carbon cache.
+    datapoints = self.pop(metric)
+
+    # retain last N datapoints in carbon cache,
+    # let's push them back
+    for i in range(1, settings.MAX_RETAINED_LATEST_DATAPOINTS + 1):
+      if i > len(datapoints):
+        break
+      timestamp, tup = datapoints[-i]
+      value, is_flushed = tup
+      self.store(metric, (timestamp, value), is_flushed=True)
+
+    # Actions before returning
+    # 1. filter out datapoints that have been already flushed.
+    # 2. strip out is_flushed field, as for keeping same interface as before
+    datapoints = [(timestamp, value) for (timestamp, (value, is_flushed)) in datapoints if not is_flushed]
+
+    # Update total_unflushed
+    with self.lock:
+      self.total_unflushed -= len(datapoints)
+
+    return (metric, datapoints)
 
   def get_datapoints(self, metric):
     """Return a list of currently cached datapoints sorted by timestamp"""
-    return sorted(self.get(metric, {}).items(), key=by_timestamp)
+    # let's keep old interface, strip out is_flushed before return
+
+    datapoints = sorted(self.get(metric, {}).items(), key=by_timestamp)
+    return [(timestamp, value) for (timestamp, (value, is_flushed)) in datapoints]
```

Comment:
probably unnecessary, but you could call sorted on get_unsorted_datapoints here. some DRY.